File: python/Reverse Linked List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseList(self, head):
        if not head:
            return None

        values = []
        while head:
            values.append(head.val)
            head = head.next
        
        values.reverse()
        
        new_list_head = ListNode(values[0])
        cursor = new_list_head

        for i in range(1, len(values)):
            cursor.next = ListNode(values[i])
            cursor = cursor.next
            
        return new_list_head

File: cpp/Reverse Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (head == NULL || head->next == NULL) {
            return head;
        }
        ListNode *dummy = new ListNode(-1);
        ListNode *p = head, *q = head, *r = head->next;
        
        while (r) {
            q = r;
            r = r->next;
            q->next = p;
            p = q;
        }
        head->next = NULL;
        return p;
    }
};
